### PR TITLE
[inductor] early memory reuse and new operators

### PR DESCRIPTION
**Pull request by jansel**
There is still a bug in the inplacing logic I need to track down.

### CODE REVIEW

File: torchinductor/lowering.py [hunk 1]
```diff
@@ -22,6 +22,37 @@
 lowerings = {}
 aten = torch.ops.aten
 
+# TODO(jansel): figure out a way generate this automatically
+#               these come as dtype args to kernels
+# based on https://github.com/pytorch/pytorch/blob/9e3eb329df8f701/c10/core/ScalarType.h#L28
+DTYPE_ID_LOOKUP = {
```
Comment (Chillee) [inline torchinductor/lowering.py:28]: lol, cc: @ezyang

Comment (ezyang) [reply torchinductor/lowering.py:28]: This is fixed on master pytorch

**Review by bertmaher — APPROVED** [review]
lgtm

File: torchinductor/lowering.py (hunk 1 above)
Comment (jansel) [reply torchinductor/lowering.py:28]: @ezyang I definitely still need this on the latest pytorch master.  Example of an op getting `dtype=6` is `aten.new_ones`, but there are many others.

Comment (ezyang) [reply torchinductor/lowering.py:28]: oh I bet this is because optionals are not being handled correctly

**Pull request merged**